central/scannerdefinitions/handler/handler.go

```diff
@@ -316,3 +306,99 @@ func (h *httpHandler) cleanupUpdaters(cleanupAge time.Duration) {
 		}
 	}
 }
+
+// openMostRecentDefinitions opens the latest Scanner Definitions based on
+// modification time. It's either the one selected by `uuid` if present and
+// online, otherwise fallback to the manually uploaded definitions. The file
+// object can be `nil` if the definitions file does not exist, rather than
+// returning an error.
+func (h *httpHandler) openMostRecentDefinitions(uuid string) (file *os.File, modTime time.Time, err error) {
+	// If in offline mode or uuid is not provided, default to the offline file.
+	if !h.online || uuid == "" {
+		file, modTime, err = h.offlineFile.Open()
+		return
+	}
+
+	// Start the updater, can be called multiple times for the same uuid, but will
+	// only start the updater once. The Start() call blocks if the definitions was
+	// not downloaded yet.
+	u := h.getUpdater(uuid)
+	u.Start()
+
+	// Open both the "online" and "offline", and save their modification times.
+	onlineFile, onlineTime, err := u.file.Open()
+	if err != nil {
+		return
+	}
+	offlineFile, offlineTime, err := h.offlineFile.Open()
+	if err != nil {
+		utils.IgnoreError(onlineFile.Close)
+		return
+	}
+
+	// Return the most recent file, notice that if both don't exist, nil is returned
+	// since modification time will be zero.
+
+	if offlineTime.After(onlineTime) {
+		file, modTime = offlineFile, offlineTime
+		utils.IgnoreError(onlineFile.Close)
+	} else {
+		file, modTime = onlineFile, onlineTime
+		utils.IgnoreError(offlineFile.Close)
+	}
+
+	return
+}
+
+// openFromArchive returns a file object for a name within the definitions
+// bundle. The file object has a file descriptor allocated on the filesystem, but
+// its name is removed. Meaning once the file object is closed, the data will be
+// freed in filesystem by the OS.
+func openFromArchive(archiveFile string, fileName string) (*os.File, error) {
+	// Open zip archive and extract the fileName.
+	zipReader, err := zip.OpenReader(archiveFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "opening zip archive")
+	}
+	defer utils.IgnoreError(zipReader.Close)
+	fileReader, err := zipReader.Open(fileName)
+	if err != nil {
+		return nil, errors.Wrap(err, "extracting")
+	}
+	defer utils.IgnoreError(fileReader.Close)
+
+	// Create a temporary file and remove it, keeping the file descriptor.
+	tmpDir, err := os.MkdirTemp("", definitionsBaseDir)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating temporary directory")
+	}
+	tmpFile, err := os.Create(filepath.Join(tmpDir, path.Base(fileName)))
+	if err != nil {
+		// Best effort to clean.
+		_ = os.RemoveAll(tmpDir)
```

Comment:
missed this before, but I think we should just `defer os.RemoveAll(tmpDir)` once we know `err != nil` for `os.MkdirTemp`. I see below we return the error if `os.RemoveAll` fails. Should we block the handler if removal fails, though? Maybe just log the error?

Reply:
> missed this before, but I think we should just defer os.RemoveAll(tmpDir) once we know err != nil for os.MkdirTemp.

Why?

> I see below we return the error if os.RemoveAll fails. Should we block the handler if removal fails, though? Maybe just log the error?

This is *very unlikely* to happen unless there is a bug in the code or the node/container is failing. If anything, I think we should crash if we cannot create or delete directories.

But in any case, I don't want to leave files in the filesystem with data if something fails during or after the data was copied. As the code and comment is explicitly stating, we create the file and remove it, and don't sync() the data.

Reply:
We want to do `os.RemoveAll()` no matter what, right? Doing `defer os.RemoveAll()` after

```
// Create a temporary file and remove it, keeping the file descriptor.
tmpDir, err := os.MkdirTemp("", definitionsBaseDir)
if err != nil {
	return nil, errors.Wrap(err, "creating temporary directory")
}
```

ensures it is run no matter the path. I don't think it matters if we remove the directory before or after the copy below